### PR TITLE
Add integration tests for cloud storage

### PR DESCRIPTION
**Pull request by c-w**
In order to facilitate onboarding of new storage backends in the future (e.g. Azure Blob Storage), this pull request refactors the existing tests for the local storage backend to also run for cloud storage backends.

The integration tests are disabled by default. For Google Cloud Storage, the integration tests can be enabled by setting the environment variables `GCP_STORAGE_KEY` and `GCP_STORAGE_SECRET`.

This pull request also fixes two edge-cases in the storage implementation that were uncovered by the tests:

1) The use of `datetime.strftime('%s')` relies on platform-specific undocumented behavior and may not work across Python versions or deployment environments. This was switched to a cross-platform way of converting a datetime to epoch via `datetime.timestamp()`.

2) Submitting malicious files to Google Cloud Storage returned a 400 error rejecting the file. To mirror the behavior of the local storage provider, "special" parts of files uploaded to non-local storage such as `.` and `..` now get sanitized to underscores.

Furthermore, the pull request also:

3) Updates to the latest version of apache-libcloud which improves handling of missing files. Without this update, fetching non-existing blobs produces a 302 result and `test_malicious_local_get_blob` fails.

4) Adds new tests to verify the results of binary downloads (previously only text downloads were verified).

5) Disables parallel test execution on CircleCI. The cloud storage tests rely on mutating environment variables before creation of the Flask app which means that the tests are no longer guaranteed to be threadsafe.

All the changes described above can be found in d16b5ac. The commit depends on 42e40d6 which introduces more parameterization to the test settings such that the storage backend used during the tests can be configured via environment variables.

The pull request also contains a number of additional minor fixes:
- 3b273be and 0ff08d0: Remove creation of directories in settings module and instead ensure that the storage module always creates a container if it doesn't yet exist.
- 0559126: Add SQLite journal file to git-ignore to avoid committing the binary file (e.g during long tests runs).
- aff57b6: Remove direct usage of pytest for exception checking and instead use standard unittest like the rest of the tests in the project.

### CODE REVIEW

File: tests/test_files_cloud.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    secret_env_name = ""
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+
+        cls.storage_key = os.environ.get(cls.key_env_name)
+        cls.storage_secret = os.environ.get(cls.secret_env_name)
+        if not cls.storage_key or not cls.storage_secret:
+            raise unittest.SkipTest("Cloud storage credentials for "
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Done in 7b2040c

File: tests/test_files_cloud.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+"""
+This module contains integration tests which verify the behavior of the storage
+module against supported cloud storage providers. The tests for the local
+(directory-based) storage backend are re-used and run for each supported cloud
+storage provider. To add tests for a new provider, create a new subclass of
+CloudTestFile.
+
+To run tests for Google Cloud Platform, set the following environment variables:
+- GCP_STORAGE_KEY
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Added TODO note in 66dfffb

File: server/storage.py [hunk 3]
```diff
@@ -61,7 +61,10 @@ def init_app(self, app):
 
         self.driver = driver_cls(key, secret)
         # Also test credentials by getting the container
-        self.container = self.driver.get_container(container_name=self.container_name)
+        try:
+            self.container = self.driver.get_container(self.container_name)
+        except ContainerDoesNotExistError:
+            self.container = self.driver.create_container(self.container_name)
```
[the remark this answers is not in the document]
Comment (c-w) [reply server/storage.py]: Added TODO note in 66dfffb

File: tests/test_files_cloud.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        cls.storage_secret = os.environ.get(cls.secret_env_name)
+        if not cls.storage_key or not cls.storage_secret:
+            raise unittest.SkipTest("Cloud storage credentials for "
+                                    "%s not configured" % cls.storage_provider)
+
+    def create_app(self):
+        os.environ["STORAGE_PROVIDER"] = self.storage_provider
+        os.environ["STORAGE_KEY"] = self.storage_key
+        os.environ["STORAGE_SECRET"] = self.storage_secret
+        os.environ.setdefault("STORAGE_CONTAINER", "okpycloudfilestest%d" % random.randint(0, 100000))
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Done in 7b2040c

File: tests/test_files_cloud.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        cls.storage_key = os.environ.get(cls.key_env_name)
+        cls.storage_secret = os.environ.get(cls.secret_env_name)
+        if not cls.storage_key or not cls.storage_secret:
+            raise unittest.SkipTest("Cloud storage credentials for "
+                                    "%s not configured" % cls.storage_provider)
+
+    def create_app(self):
+        os.environ["STORAGE_PROVIDER"] = self.storage_provider
+        os.environ["STORAGE_KEY"] = self.storage_key
+        os.environ["STORAGE_SECRET"] = self.storage_secret
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Added TODO note in 66dfffb

File: tests/test_files_cloud.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+        return super().create_app()
+
+    def tearDown(self):
+        super().tearDown()
+
+        del os.environ["STORAGE_PROVIDER"]
+        del os.environ["STORAGE_KEY"]
+        del os.environ["STORAGE_SECRET"]
+        del os.environ["STORAGE_CONTAINER"]
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Refactored the environment variable handling in 3afc998. We now only set the environment variables at class creation time and restore them back to original values at class cleanup time.

File: tests/test_files_cloud.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+        for obj in storage.container.list_objects():
+            obj.delete()
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Done in f95d8f8

File: tests/test_files_cloud.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        storage.container.delete()
```
[the remark this answers is not in the document]
Comment (c-w) [reply tests/test_files_cloud.py]: Done in f95d8f8

**Comment by colinschoen** [comment]
Thanks Clemens! Taking a look now.

File: tests/test_files_cloud.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        storage_key = os.environ.get(cls.key_env_name)
+        storage_secret = os.environ.get(cls.secret_env_name)
+
+        if not storage_key or not storage_secret:
+            raise unittest.SkipTest("Cloud storage credentials for {} not configured".format(cls.storage_provider))
+
+        cls.set_environment_variable("STORAGE_PROVIDER", cls.storage_provider)
+        cls.set_environment_variable("STORAGE_KEY", storage_key)
+        cls.set_environment_variable("STORAGE_SECRET", storage_secret)
+        cls.set_environment_variable("STORAGE_CONTAINER", cls._storage_container)
```
Comment (colinschoen) [inline tests/test_files_cloud.py]: I think this will prevent users from setting a storage container to use for the integration test via an environment variable. Can we first check to see if `STORAGE_CONTAINER` is set. If not then use `cls._storage_container`.

Comment (c-w) [reply tests/test_files_cloud.py]: I believe this is by design. The way that the code is currently designed is that the touch-point to enable/disable certain integration tests are the environment variables in key_env_name and secret_env_name. Being an abstract test base, the CloudTestFile never actually executes and each cloud-specific subclass will set the storage provider environment variable that it's testing for its tests and then we revert back to the default test storage (local) for the remainder of the tests suite.

For example, if the GCP_STORAGE_KEY and GCP_STORAGE_SECRET environment variables are set, the GoogleCloudTestFile sets the storage provider to GOOGLE_STORAGE for the 14 tests it inherits from TestFile and then we revert back to local for the rest of the test suite.

```py
class GoogleCloudTestFile(CloudTestFile):
    storage_provider = "GOOGLE_STORAGE"
    key_env_name = "GCP_STORAGE_KEY"
    secret_env_name = "GCP_STORAGE_SECRET"

    test_prefix_expected_obj_name = 'test/fizz.txt'
    test_malicious_directory_traversal_expected_obj_name = 'test/_/_/fizz.txt'
```

Similarly, if we wanted to make a additional integration test for S3 storage, we'd add a new class like this:

```py
class AWSCloudTestFile(CloudTestFile):
    storage_provider = "S3"
    key_env_name = "S3_STORAGE_KEY"
    secret_env_name = "S3_STORAGE_SECRET"

    test_prefix_expected_obj_name = 'test/fizz.txt'
    test_malicious_directory_traversal_expected_obj_name = 'test/_/_/fizz.txt'
```

This setup of the code enables us to run integration tests for the 14 storage tests against multiple backends (local, GCP, AWS, etc.) during the same test run without having to re-run the entire test suite (100+ tests) for each of the storage backends.

Hope that clears things up (and I'd be happy to hear suggestions on how this could be made clearer in the code to avoid future confusion).

If we want to run the entire test suite for each of the storage providers (instead of just the TestFile cases), let me know and I can make that change.

Comment (colinschoen) [reply tests/test_files_cloud.py]: Maybe I am misunderstanding, but I am referring to `STORAGE_CONTAINER` not `STORAGE_PROVIDER`.

Comment (c-w) [reply tests/test_files_cloud.py]: Apologies, I completely misread that for some reason. Implemented respecting of any pre-existing storage container value in 8799fcd.

File: tests/test_files_cloud.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    @classmethod
+    def tearDownClass(cls):
+        super().tearDownClass()
+
+        cls.restore_environment_variable("STORAGE_PROVIDER")
+        cls.restore_environment_variable("STORAGE_KEY")
+        cls.restore_environment_variable("STORAGE_SECRET")
+        cls.restore_environment_variable("STORAGE_CONTAINER")
+
+        container = storage.driver.get_container(cls._storage_container)
```
Comment (colinschoen) [inline tests/test_files_cloud.py]: This will throw a `ContainerDoesNotExistError` when `STORAGE_CONTAINER` is set by the user.

Comment (c-w) [reply tests/test_files_cloud.py]: Thanks for the continued thorough review, @colinschoen.

At this point, I think it's simpler if we just require the storage container to be already created by the time that the integration tests run, so I've implemented that in f1b2e38. We can re-use the same container for all test runs so it's only a one-time setup cost.

Requiring a pre-existing container simplifies handling of the environment variables and also means that we don't require a special service account with create/delete bucket permissions to run the tests.

Note that I've introduced a new environment variable for the storage account name used in the integration tests so that a container name for the local tests can be set via `CONTAINER_NAME` and a different container name for the the GCP cloud storage integration test can be set via `GCP_STORAGE_CONTAINER` (to prevent potential clashes between names that are valid local containers but not in the cloud provider).

Here is how I ran the tests:

1. Create a new Storage project in GCP.
2. Set the project as the default project for interoperable access.
3. Copy the access key to `GCP_STORAGE_KEY` and the secret to `GCP_STORAGE_SECRET`.
4. Create a new bucket and copy its name to `GCP_STORAGE_CONTAINER`.
5. Execute `python manage.py test` (passes).
6. Unset all the `GCP_` environment variables.
7. Run `python manage.py test` (passes).

Thoughts?

Comment (colinschoen) [reply tests/test_files_cloud.py]: Awesome, I like this a lot. I will give it a go.

Comment (colinschoen) [reply tests/test_files_cloud.py]: Haven't had a chance to debug what is going on, but I am failing a few `GoogleCloudTestFile` tests as well as `TestFile` tests when setting the following environment variables: `GCP_STORAGE_KEY`, `GCP_STORAGE_SECRET`, `GCP_STORAGE_CONTAINER`. 

Interestingly, I don't fail those `TestFile` tests when the said environment variables are unset. 

I will be taking a look to figure out why. 
[test_output.txt](https://github.com/Cal-CS-61A-Staff/ok/files/1940632/test_output.txt)

**Comment by c-w** [comment]
The CI failed due to `TestApi.test_get_backup` which compares date-time values at seconds granularity ([Travis CI link](https://travis-ci.org/Cal-CS-61A-Staff/ok/jobs/373293928)). Running the test locally passes so it's a flaky one. Rebasing to re-trigger the CI run.

The test could be fixed by comparing date-time values with a certain tolerance threshold (e.g. one second) but that's outside of the scope of this PR.

**Comment by c-w** [comment]
CI is now green. Let me know if you'd like me to do anything else before merging, @colinschoen (And thanks for your continued support throughout the pull request!)

**Comment by colinschoen** [comment]
Thank you and the rest of the Microsoft folks for your contributions!!
